rosplan_knowledge_base/src/VALVisitorProblem.cpp

```diff
@@ -15,26 +16,54 @@ namespace KCL_rosplan {
     /*--------------*/
 
     /**
-     * Visit an prop to pack into ROS message
+     * Visit a proposition to pack into ROS message
      */
     void VALVisitorProblem::visit_proposition(VAL::proposition *p) {
 
         last_prop.typed_parameters.clear();
+        last_prop.labels.clear();
 
-        // predicate name
-        rosplan_knowledge_msgs::DomainFormula formula;
-        last_prop.name = p->head->symbol::getName();
-        // predicate variables
+        last_prop.name = p->head->getName();
 
+        std::vector<std::string> predicateLabels;
 
+        // parse domain for predicates
+        VAL::pred_decl_list *predicates = domain->predicates;
+        for (VAL::pred_decl_list::const_iterator predicateIterator = predicates->begin(); predicateIterator != predicates->end(); predicateIterator++) {
+
+            VAL::pred_decl *tempPredicate = *predicateIterator;
+
+            // compare pedicate name with last proposition name
+            if (tempPredicate->getPred()->symbol::getName() == last_prop.name) {
+
+                // iterate the predicate symbols
+                for (VAL::var_symbol_list::const_iterator varSymbolIterator = tempPredicate->getArgs()->begin();
+                     varSymbolIterator != tempPredicate->getArgs()->end(); varSymbolIterator++) {
+
+                    // add labels to predicateLabels
+                    const VAL::var_symbol *tempVarSymbol = *varSymbolIterator;
+                    predicateLabels.push_back(tempVarSymbol->pddl_typed_symbol::getName());
+                }
+            }
+        }
+
+        int index = 0;
         for (VAL::parameter_symbol_list::const_iterator vi = p->args->begin(); vi != p->args->end(); vi++) {
             const VAL::parameter_symbol* var = *vi;
             diagnostic_msgs::KeyValue param;
             param.key = var->pddl_typed_symbol::getName();
```

Comment:
This is the type, (I think) do we need to store this here?

In the KB, the facts are stored only with predicate name, and a KeyValue array of Label->Object.